docs/decisions/0015-expiration-improvements.rst

```diff
@@ -0,0 +1,94 @@
+0015 Expiration and Cancellation Data Improvements
+**************************************************
+
+Status
+======
+Proposed - December 2023
+
+Context
+=======
+There are subtleties around the Assignment cancellation and expiration user
+experience in the enterprise learner-portal frontend that require a more nuanced
+and descriptive approach to the data we persist around these states, and
+how that data is represented in API response payloads for consumption
+by the learner-portal frontend.
+
+How expiration currently works
+------------------------------
+There's currently a management command (run via a cron) that toggles the state
+of Assignments to ``CANCELLED`` under the following conditions:
+
+1. The current date is more than 90 days after the *creation* date of the assignment. We use the creation
+   date as a proxy for the date at which an assignment was allocated, which is problematic under
+   the edge case where an assignment is allocated, cancelled, and re-allocated later.
+2. The current date is greater than the inferred enrollment deadline for the assigned course.
+3. The current date is greater than the expiration date of the subsidy associated with the Assignment's
+   access policy record.
+
+Learners are notified of expiration and cancellation via the learner-portal frontend
+------------------------------------------------------------------------------------
+The learner-portal notifies learners of assignment cancellation and expiration.  These notifications
+are required to be dismissable, which means we need to persist state about dismissal
+for assignments in the backend.
+
+Decision
+========
+
+``EXPIRED`` should be a distinct Assignment State
+-------------------------------------------------
+This will help disambiguate behavior and simplify/remove business logic
+from the learner-portal frontend.  We may wish to also store the expiration reason
+for an assignment (i.e. which of the 3 cases above caused the assignment to become expired).
+
+Persist the allocation, cancellation, and expiration dates
+----------------------------------------------------------
+This also helps simplify and clarify the true state of an assignment for
+consumption by the learner-portal frontend.
+
+Cancellation/Expiration acknowledgement should be an action
+-----------------------------------------------------------
+There should be an action type to support this, and an endpoint scoped to the assignment identifier
+to create such action records.  This will greatly simplify logic in the learner-portal frontend
```

Comment:
The dismiss action deals with one _or more_ assignments, across all assignment configurations for the customer. Ideally, the frontend would have a single API call (or least one API call per assignment configuration) in order to dismiss relevant canceled/expired assignments. As is, this implies the frontend would need to make an API call per assignment, which adds complexity and frontend performance concerns (e.g., all N API requests would need to be resolved in parallel, have a reasonable interaction for error handling should one or more of those requests fail, etc. resolve before the alert or assignments can be dismissed.

Edit: I see you mentioned this below:

```
It's ok to not have a bulk API view to acknowledge multiple assignment expiration in a single request -
these acknowledgements are one-time activities for most assignments, and we can live with the cost
of making multiple requests from the frontend for now.  In return, we get to deal with a simpler API view
and logic.
```

What are your concerns around at least making it one API request per assignment configuration (given the API schema today for assignments is specific to assignment configurations)? Given we don't really have any inclination yet for how cancelled/expired assignments learners will have in practice, I'm still a bit uncomfortable about making 1 API request per assignment.

Reply:
> 1 API request per assignment configuration
That's reasonable, I'll update!